Adding functionality to include data information

liamegan opened the pull request:
Added a simple component to look for an index.js file in a relative `data` folder. This file should export a promise, for example:

```
module.exports = async function() {
  const data = {
    testing: 'a'
  };
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(data);
    }, 300);
  });
}
```

This data will then be resolved to an object called `models` available to PUG.

liamegan committed:
Adding functionality to include data information

ste-vg commented:
Would it be better to have it look for any data/*.js and data/*.json and load them all, not just the index? The data could be added to the `models` object with the filename. For example `models.[filename].[property]`.

liamegan commented:
We could do that for sure, and I thought of that as well, however I opted
for this approach because it's the unopinionated route.

For example if that's how you wanted to structure your model you could
easily do so by writing a bunch of files, collecting them with your index
and structuring the model like that, but this doesn't mandate that as being
necessary. The advantage with this approach is that it's just JavaScript,
so we can do whatever we want.

I'm happy to be convinced otherwise, but I ce to the conclusion here that
the less opinionated approach was probably better.

On Thu., Jun. 3, 2021, 02:30 Steve Gardner, ***@***.***>
wrote:

> Would it be better to have it look for any data/*.js and data/*.json and
> load them all, not just the index? The data could be added to the models
> object with the filename. For example models.[filename].[property].
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/wethegit/sweet-potato/pull/20#issuecomment-853729935>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEZJ2G6P4XJ3MROET3IBCTTQ5DTVANCNFSM457XMA5A>
> .
>

marlonmarcello commented on cooker/bin/commands/start/request-listener.js:
```diff
@@ -144,7 +144,7 @@ function _respond(
  */
 function _doesntExist(res, file) {
-  const prettyPath = path.relative(config.CWD, file);
+  const prettyPath = path.relative('/', file);
```
Was this broken @liamegan ?

liamegan replied:
It wasn't broken, but the site I chose to test this on (Loz) was using the old name for the config file which was causing all sorts of problems. This was my attempt to test it (I was trying to test whether the SP thought it was in the src folder or not), it prints out the full location of the file, rather than the file relative to the CWD, which doesn't yield relevant information if you're wondering about things above the CWD.

Either way, this shouldn't really have been included in the PR and I'll roll it back.

marlonmarcello commented on cooker/lib/pages.js:
```diff
@@ -158,6 +159,26 @@ async function getDataFromYaml(file) {
   return result;
 }
 
+async function getDataFromDataInclude(file) {
```
Let's add some JSDocs in here.

liamegan replied:
Yes! I meant to do that.

marlonmarcello reviewed (changes requested):
Other than the two very minor comments, this is looking good.

liamegan committed:
Supplying the working folder as a parameter of the model function.

This allows us to run things like file collection within the data folder without opinions of where it's comeing from.

liamegan committed:
Adding documentation to getDataFromDataInclude

liamegan committed:
Adding relative path back into doesn't exist error messaging

liamegan commented:
Alright, updates made.

@ste-vg

Given this file layout:
![image](https://user-images.githubusercontent.com/627944/120667411-518cce00-c442-11eb-853d-efe4ccd5f226.png)

And this setup, I could load these yaml files just by using a bit of extra javascript. My index.js looks like this:

```javascript
const yaml = require("js-yaml");
const fse = require("fs-extra");
const path = require('path');

async function readYaml(file) {
  let result = {};

  if (!fse.pathExistsSync(file)) return result;

  try {
    const content = await fse.readFile(file, "utf8");
    result = yaml.load(content);
  } catch (error) {
    logger.error(
      [`Can't compile yaml`, file],
      error
    );

    throw error;
  }

  return result;
}

module.exports = async function(relativepath) {
  const data = {
    testing: 'a'
  };

  const datafiles = {};
  await fse.readdir(relativepath)
    .then(filenames => {
      for(let i = 0; i < filenames.length; i++) {
        const filename = filenames[i];
        if(path.extname(filename) === '.yaml') {
          datafiles[filename.split('.')[0]] = path.join(relativepath, filename);
        }
      }
    });
  for(let i in datafiles) {
    data[i] = await readYaml(datafiles[i]);
  }

  return data;
}
```

Yes it places more effort onto the developer and we may end up rolling this functionality into the SP in future (if it becomes the dominant pattern), however I like the unopinionated approach for now. Let me know if you disagree @marlonmarcello or @ste-vg (or anyone). I'm more than happy to be wrong about this.

ste-vg commented:
> We could do that for sure, and I thought of that as well, however I opted for this approach because it's the unopinionated route. For example if that's how you wanted to structure your model you could easily do so by writing a bunch of files, collecting them with your index and structuring the model like that, but this doesn't mandate that as being necessary. The advantage with this approach is that it's just JavaScript, so we can do whatever we want. I'm happy to be convinced otherwise, but I ce to the conclusion here that the less opinionated approach was probably better.

Not sure I agree. There's nothing stopping you from putting other data in another folder and pulling it in via the data/index.js if that's what you want to do, while also supporting all files in data automatically get pulled in. Not unlike how all pug and scss files in the pages folder automatically get built.

It's just a personal preference really but I think I'd prefer just being able to put some JSON in the data folder and not worry about making sure it's manually pulled into the index.js 

Not a big deal either way.
